Catch integer overflow on 32-bits in JSWebAssemblyArray.h

aoikonomopoulos opened the pull request:
#### 4b568b921b7a39476813832bace8b169d47d751b
<pre>
Catch integer overflow on 32-bits in JSWebAssemblyArray.h
<a href="https://bugs.webkit.org/show_bug.cgi?id=289523">https://bugs.webkit.org/show_bug.cgi?id=289523</a>

Reviewed by Yusuke Suzuki.

Array size times element size may wrap around when size_t is only 4
bytes. Catch and propagate.

* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::JSC_DEFINE_NOEXCEPT_JIT_OPERATION):
* Source/JavaScriptCore/wasm/WasmOperationsInlines.h:
(JSC::Wasm::tryFillArray):
(JSC::Wasm::arrayNew):
(JSC::Wasm::tryCopyElementsInReverse):
(JSC::Wasm::arrayNewFixed):
(JSC::Wasm::createArrayFromDataSegment):
(JSC::Wasm::arrayNewElem):
(JSC::Wasm::fillArray): Deleted.
(JSC::Wasm::copyElementsInReverse): Deleted.
* Source/JavaScriptCore/wasm/js/JSWebAssemblyArray.h:

Canonical link: <a href="https://commits.webkit.org/292355@main">https://commits.webkit.org/292355@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/65b698f7a6f5b06d6973805fb7abe038724478f4

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/95783 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15393 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/5312 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/100837 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/46288 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/15680 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/23824 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/73046 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30284 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/98786 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/11741 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86508 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53375 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11468 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/4243 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/45627 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/88451 "Built successfully and passed tests") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/81652 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/4350 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/102868 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/94403 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/22842 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/16675 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/82079 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/23094 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82539 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81440 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20397 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/26020 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/137/builds/3478 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/16178 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/22809 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/27970 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/117882 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22468 "Built successfully") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/33373 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/25944 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/24210 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

Constellation commented on Source/JavaScriptCore/wasm/js/JSWebAssemblyArray.h:
```diff
@@ -58,7 +58,12 @@ class JSWebAssemblyArray final : public WebAssemblyGCObjectBase {
 
     static JSWebAssemblyArray* create(VM& vm, Structure* structure, Wasm::FieldType elementType, unsigned size, RefPtr<const Wasm::RTT>&& rtt)
```
If it fails, function name should be tryCreate.

aoikonomopoulos replied:
Oops, thanks.

Constellation commented on Source/JavaScriptCore/wasm/WasmOperationsInlines.h:
```diff
@@ -292,6 +300,8 @@ inline EncodedJSValue arrayNewElem(JSWebAssemblyInstance* instance, uint32_t typ
     JSGlobalObject* globalObject = instance->globalObject();
     VM& vm = globalObject->vm();
     auto* array = JSWebAssemblyArray::create(vm, globalObject->webAssemblyArrayStructure(), FieldType { storageType, Mutability::Mutable }, arraySize, WTFMove(arrayRTT));
+    if (!array)
```
UNLIKELY

aoikonomopoulos replied:
Thanks, wrapping the values in those tests in UNLIKELY() .

Constellation commented on Source/JavaScriptCore/wasm/WasmOperationsInlines.h:
```diff
@@ -201,6 +207,8 @@ EncodedJSValue createArrayFromDataSegment(JSWebAssemblyInstance* instance, Field
     JSGlobalObject* globalObject = instance->globalObject();
     VM& vm = globalObject->vm();
     auto* array = JSWebAssemblyArray::create(vm, globalObject->webAssemblyArrayStructure(), elementType, arraySize, WTFMove(rtt));
+    if (!array)
```
UNLIKELY

Constellation commented on Source/JavaScriptCore/wasm/WasmOperationsInlines.h:
```diff
@@ -179,6 +183,8 @@ inline JSValue arrayNewFixed(JSWebAssemblyInstance* instance, uint32_t typeIndex
     default:
         RELEASE_ASSERT_NOT_REACHED();
     }
+    if (!array)
```
UNLIKELY

Constellation commented on Source/JavaScriptCore/wasm/WasmOperationsInlines.h:
```diff
@@ -132,7 +136,7 @@ JSWebAssemblyArray* copyElementsInReverse(JSWebAssemblyInstance* instance, Wasm:
     VM& vm = instance->vm();
 
     auto* array = JSWebAssemblyArray::create(vm, globalObject->webAssemblyArrayStructure(), fieldType, size, WTFMove(rtt));
```
We should add UNLIKELY(!array) if statement separately.

Constellation commented on Source/JavaScriptCore/wasm/WasmOperationsInlines.h:
```diff
@@ -121,6 +123,8 @@ inline JSValue arrayNew(JSWebAssemblyInstance* instance, uint32_t typeIndex, uin
         return jsNull();
 
     auto* array = JSWebAssemblyArray::create(vm, globalObject->webAssemblyArrayStructure(), fieldType, size, WTFMove(rtt));
+    if (!array)
```
UNLIKELY

Constellation commented on Source/JavaScriptCore/wasm/WasmOperationsInlines.h:
```diff
@@ -99,7 +100,8 @@ inline JSValue arrayNew(JSWebAssemblyInstance* instance, uint32_t typeIndex, uin
         RELEASE_ASSERT_NOT_REACHED();
     }
-    ASSERT(array);
+    if (!array)
```
UNLIKELY

Constellation commented on Source/JavaScriptCore/wasm/WasmOperationsInlines.h:
```diff
@@ -59,7 +59,8 @@ JSWebAssemblyArray* fillArray(JSWebAssemblyInstance* instance, Wasm::FieldType f
 
     auto* array = JSWebAssemblyArray::create(vm, globalObject->webAssemblyArrayStructure(), fieldType, size, WTFMove(rtt));
-    array->fill(vm, 0, static_cast<T>(value), size);
+    if (array)
```
UNLIKELY

Constellation commented on Source/JavaScriptCore/wasm/WasmOperationsInlines.h:
```diff
@@ -292,6 +300,8 @@ inline EncodedJSValue arrayNewElem(JSWebAssemblyInstance* instance, uint32_t typ
     VM& vm = globalObject->vm();
     auto* array = JSWebAssemblyArray::create(vm, globalObject->webAssemblyArrayStructure(), FieldType { storageType, Mutability::Mutable }, arraySize, WTFMove(arrayRTT));
+    if (!array)
+        return JSValue::encode(jsNull());
```
Is returning null correct? What is the type of wasm instruction we are expecting?

aoikonomopoulos replied:
We're expecting `array.new_elem` from the GC extension [spec](https://webassembly.github.io/gc/core/appendix/index-instructions.html#index-instr). AFAICT our callers convert an `isNull()` result to `Wasm::ExceptionType::BadArrayNewInitElem`, for which the associated message is "Out of bounds or failed to allocate in array.new_elem", so `jsNull()` seems appropriate here (I guess if we were drawing the distinction, this would be a "failed to allocate").

eugeneia commented on Source/JavaScriptCore/wasm/WasmOperationsInlines.h:
```diff
@@ -58,8 +58,9 @@ JSWebAssemblyArray* fillArray(JSWebAssemblyInstance* instance, Wasm::FieldType f
-    auto* array = JSWebAssemblyArray::create(vm, globalObject->webAssemblyArrayStructure(), fieldType, size, WTFMove(rtt));
-    array->fill(vm, 0, static_cast<T>(value), size);
+    auto* array = JSWebAssemblyArray::tryCreate(vm, globalObject->webAssemblyArrayStructure(), fieldType, size, WTFMove(rtt));
+    if (UNLIKELY(array))
```
This seems off: !array is the UNLIKELY case, no?

aoikonomopoulos replied:
Yes, thanks for catching this!

sosukesuzuki commented on Source/JavaScriptCore/wasm/WasmOperationsInlines.h:
```diff
@@ -58,8 +58,9 @@ JSWebAssemblyArray* fillArray(JSWebAssemblyInstance* instance, Wasm::FieldType f
 
-    auto* array = JSWebAssemblyArray::create(vm, globalObject->webAssemblyArrayStructure(), fieldType, size, WTFMove(rtt));
-    array->fill(vm, 0, static_cast<T>(value), size);
+    auto* array = JSWebAssemblyArray::tryCreate(vm, globalObject->webAssemblyArrayStructure(), fieldType, size, WTFMove(rtt));
```
Should we rename `fillArray` to `tryFillArray`?

aoikonomopoulos replied:
Makes sense, thanks!

sosukesuzuki commented on Source/JavaScriptCore/wasm/WasmOperationsInlines.h:
```diff
@@ -131,7 +135,9 @@ JSWebAssemblyArray* copyElementsInReverse(JSWebAssemblyInstance* instance, Wasm:
     VM& vm = instance->vm();
 
-    auto* array = JSWebAssemblyArray::create(vm, globalObject->webAssemblyArrayStructure(), fieldType, size, WTFMove(rtt));
+    auto* array = JSWebAssemblyArray::tryCreate(vm, globalObject->webAssemblyArrayStructure(), fieldType, size, WTFMove(rtt));
```
Should we rename `copyElementsInReverse` to `tryCopyElementsInReverse` ?

aoikonomopoulos replied:
Ditto.

aoikonomopoulos committed:
Catch integer overflow on 32-bits in JSWebAssemblyArray.h
https://bugs.webkit.org/show_bug.cgi?id=289523

Reviewed by Yusuke Suzuki.

Array size times element size may wrap around when size_t is only 4
bytes. Catch and propagate.

* Source/JavaScriptCore/wasm/WasmOperations.cpp:
(JSC::Wasm::JSC_DEFINE_NOEXCEPT_JIT_OPERATION):
* Source/JavaScriptCore/wasm/WasmOperationsInlines.h:
(JSC::Wasm::tryFillArray):
(JSC::Wasm::arrayNew):
(JSC::Wasm::tryCopyElementsInReverse):
(JSC::Wasm::arrayNewFixed):
(JSC::Wasm::createArrayFromDataSegment):
(JSC::Wasm::arrayNewElem):
(JSC::Wasm::fillArray): Deleted.
(JSC::Wasm::copyElementsInReverse): Deleted.
* Source/JavaScriptCore/wasm/js/JSWebAssemblyArray.h:

Canonical link: https://commits.webkit.org/292355@main
